controller/test/frame_detector/frame_detector_test.cpp

```diff
@@ -0,0 +1,144 @@
+#include "frame_detector.h"
+#include "proto_traits.h"
+#include "soft_rx_buffer.h"
+#include "gtest/gtest.h"
+
+constexpr uint32_t LEN = 20;
+
+constexpr FrameDetector<SoftRxBuffer<LEN>, LEN>::State STATE_LOST =
+    FrameDetector<SoftRxBuffer<LEN>, LEN>::State::LOST;
+constexpr FrameDetector<SoftRxBuffer<LEN>, LEN>::State STATE_WAIT_START =
+    FrameDetector<SoftRxBuffer<LEN>, LEN>::State::WAIT_START;
+constexpr FrameDetector<SoftRxBuffer<LEN>, LEN>::State STATE_RX_FRAME =
+    FrameDetector<SoftRxBuffer<LEN>, LEN>::State::RX_FRAME;
+
+constexpr uint8_t MARK = static_cast<uint8_t>('.');
+
+SoftRxBuffer<LEN> rx_buf('.');
+
+TEST(FrameDetector, TrivialRx) {
+  FrameDetector<SoftRxBuffer<LEN>, LEN> frame_detector(rx_buf);
+  EXPECT_TRUE(frame_detector.Begin());
+  ASSERT_EQ(STATE_LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(' ');
+  ASSERT_EQ(STATE_LOST, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(STATE_WAIT_START, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(STATE_RX_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(' ');
+  ASSERT_EQ(STATE_RX_FRAME, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(STATE_WAIT_START, frame_detector.get_state());
+  EXPECT_TRUE(frame_detector.is_frame_available());
```

Comment:
This is currently the only test that verifies successful reception of a frame - we need more tests of that, e.g.:
* ability to recover from all kinds of errors
* ability to receive multiple frames
* ability to receive some frames, have an error, and receive more frames
* ability to receive a frame of length almost LEN

I mean, currently the test seems to be written essentially verifying the implementation almost line-by-line, but we need tests that verify the important user-facing properties.

Is it possible e.g. to write a utility function that PutByte's a given std::string, and returns an std::vector<std::string> of received frames, and write a bunch more tests in terms of invocations of this function?

E.g. EXPECT_THAT(Unframe("foo..bar.moo...blah."), testing::ElementsAre("foo", "bar", "moo", "blah"))?

Reply:
I have started working on that, see updated tests. Will try to implement as you describe.